### PR TITLE
Adding Euclid preconditioner wrappers for ILU preconditioning options [vikramsg:master]

### PR DESCRIPTION
**Pull request by vikramsg**
This is with reference to issue https://github.com/mfem/mfem/issues/659.

I have created the wrappers and put in some default options. Example 1 is slightly modified to test the preconditioner. 

TODO
- [x] Revert changes in Example 1

### CODE REVIEW

File: linalg/hypre.hpp [hunk 1]
```diff
@@ -1134,6 +1134,28 @@ class HypreAME
...
+/// The Euclid preconditioner in hypre
+class HypreEuclid : public HypreSolver
+{
+private:
+   HYPRE_Solver euc_precond;
+
+public:
+   HypreEuclid(HypreParMatrix &A);
+
+   /// The typecast to HYPRE_Solver returns the internal sai_precond
```
[the remark this answers is not in the document]
Comment (vikramsg) [reply linalg/hypre.hpp]: Should be fixed now.

[the remark this answers is not in the document]
Comment (mlstowell) [reply linalg/hypre.hpp]: Great, thank you!

**Comment by mlstowell** [comment]
Hi, @vikramsg ,
I'm essentially ready to approve this (and I'm glad you've put this together).  The only other issue is example 1.  I know the team wants to keep this particular example as simple as possible because it's the first place most users should look.  On the other hand it would be good to have an example where we can easily switch between the Hypre preconditioners and solvers so that users can see how to experiment with different combinations (and for our own code coverage testing).

I'm curious to hear what you, @tzanio, and @v-dobrev think about this.  Should we create a new example, or perhaps a small miniapp in 'tools' that allows users to experiment with solver/preconditioner combinations on simple, user-selected, linear systems? 

Thanks,
Mark

**Comment by vikramsg** [comment]
I actually put the testing part in example 1 because that was the simplest place to test this. I don't really know where to actually place this or if it is even required, since I am not sure if there's a performance improvement with ILU for any of the examples. Personally I see a lot of improvement for my own use case which is Compressible Euler. 

I am open to whatever you guys suggest.

**Comment by tzanio** [comment]
I agree that we should revert the changes in Example 1.

I think it will be better to illustrate this in an example or miniapp where the use of Euclid really makes a (positive) difference. If we currently don't have an example like that, then I think it is fine not to include it.

**Comment by tzanio** [comment]
Approved modulus TODO items.

**Pull request merged**